### PR TITLE
Data Migration for Fideslang 2.0

### PR DESCRIPTION
**Pull request by ThomasLaPiana**
Builds on top of #3933 

### Description Of Changes

For ease of PR reviewing, this PR is focused solely on writing and testing a migration for current users to Fideslang 2.0

The following are the objects that need updating:
* [x] `Dataset` (categories on `fields`, `collections`, and top level
* [x] `Systems.PrivacyDeclarations` (`data_use`, `data_categories`, `shared_categories`)
* [x] `Systems.Ingress/Egress` (`data_categories`)
* [x] `Policy.Rules` (`data_uses`, `data_categories`)
* [x] `DataCategory/DataUse` (`fides_key`, `parent_key`)
* [x] `PrivacyNotice/PrivacyNoticeHistory/PrivacyNoticeTemplate`
* [x] `PrivacyRequest.consent_preferences`
* [x] `Consent.data_use`
* [x] `ConsentRequest.preferences`

### Code Changes

* [x] add a data migration
* [x] add a testing script for verification (`scripts/verify_fideslang_2_data_migration.py`)

### Steps to Confirm

* [ ] `nox -s dev`
* [ ] `nox -s shell`
* [ ] `fides user login`
* [ ] `fides db reset -y ; python scripts/verify_fideslang_2_data_migration.py --reload ; python scripts/verify_fideslang_2_data_migration.py  --reload` - the script will probably break on System push, so run it again to pass (seriously I'm not sure why, something with the order of how things are getting pushed I guess?)

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

**Comment by pattisdr** [comment]
Moving my last comment over from the previous PR so it doesn't get lost! 


---
Thanks for your changes @ThomasLaPiana. The migration was crashing so I added some changes here: https://github.com/ethyca/fides/commit/9308569ad9e9342d9e43256f65a3f9bd266c20cd to move this forward, but there's still more work to do. Let's all continue to test and figure out where the gaps are.

Here's some current TODO items I'm seeing after another round of testing:
- [ ]  I think this is the biggest potential issue. Migrate custom data uses and custom data categories that build off of the affected 2.0 uses and categories.
    - For example, say I have a custom data use, "user.search_history.work" that I am using on a PrivacyDeclaration. We migrate that to "user.behavior.search_history.work" on the Declaration. However, there is no "DataUse" that matches, and you get errors later when trying to update the System as a result.

- [ ] New column PrivacyDeclaration.shared_categories going out in the current release needs to be migrated now that 2.19 is out. Resurfacing https://github.com/ethyca/fides/pull/3933#discussion_r1305942158

- [ ] PrivacyNotice, PrivacyNoticeTemplate, and PrivacyNoticeHistory data_uses columns likely need to be migrated. None of our out of the box notices have data uses at risk, but customers could have added custom notices, or added data uses to existing notices.

**Comment by pattisdr** [comment]
@ThomasLaPiana @adamsachs in addition to the above, I wanted to discuss if we also need to migrate `Consent.data_use` and `PrivacyRequest.consent_preferences` (array of dictionaries, which possibly has a `data_use`)


This came up while I was hunting for additional places that use `data_use` that we are not accounting for in this migration. While we've been working on new-style consent this year, some customers are running our older version of consent, which stores saved consent preferences in the `consent` table: this has a `data_use` column. In addition, our default `config.json` file which populates the consent tab of the privacy center for old-style consent, [has `improve` as a data use](https://github.com/ethyca/fides/blob/main/clients/privacy-center/config/config.json#L79).  It is possible then, users are saving preferences against improve, which has moved to `functional`.  When we create a privacy request to propagate third party preferences, we cache the relevant `consent_preferences` on that privacy request so we have a snapshot of the preferences at that point in time, which is why an obsolete data use might be on both records. 

We may not need to modify old records.  But it seems like the most pressing thing would be to make sure whatever `config.json` customers are using was updated going forward to not have old data uses in it, which might just be customer outreach. We can talk about specifics internally.

**Comment by pattisdr** [comment]
- Similarly `consentrequest.preferences` is a jsonb stores this as well, I think this is more for record-keeping for old-style consent.  Similar to `privacyrequest.consent_preferences`.

**Comment by ThomasLaPiana** [comment]
@pattisdr can you give me some guidance on how to get at those programmatically? I'm really struggling to use the APIs, as the privacy request api needs that information for creation but doesn't return it on a `GET`

I've updated it directly in the database for `Consent.data_use`, I'm assuming that is the right move there? How can we test that?

**Comment by pattisdr** [comment]
@ThomasLaPiana

> can you give me some guidance on how to get at those programmatically? I'm really struggling to use the APIs, as the privacy request api needs that information for creation but doesn't return it on a GET

This isn't returned in the API, that was intentional! This is more of an internal detail that takes a snapshot of the preferences for the user at that point in time and propagates these to third party systems.  Your testing script might just query the privacy request directly with sqlalchemy and verify that the consent_preferences resource has been updated.

> I've updated it directly in the database for Consent.data_use, I'm assuming that is the right move there? How can we test that?

Yes, this seems right but I would expect everything to be updated directly, including the PrivacyRequest and ConsentRequest. I'm not sure what you mean by calling out this one location directly!  Testing instructions are below.  You can use the "Get previously-saved consent preferences" endpoint to check after a migration.

> (I achieved this by updating Consent.data_use, which appears to populate this?)

I'm not sure what you mean.  This is useful going forward but existing privacy requests still need to be migrated, if we're going to do that.  Also, these tables are potentially large - worth pointing this out.  This is something we shouldn't be migrating live -



# Old Consent Testing Details

## Create a Consent Request
```bash
curl -X 'POST' \
  'http://localhost:8080/api/v1/consent-request' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
   "email": "user@example.com"
}'
```
- This creates a `ConsentRequest` object, which is largely used for identity verification and tracking a particular request
- This returns a `consent_request_id` that you should use in the following request:

## Set Consent Preferences
- This creates a `Consent` object to save current preferences
- This also creates a `PrivacyRequest` to propagate preferences to third party systems and stashes the preferences on this object as well, so we know which set of preferences this request is responsible for propagating. `executable_options` in the request body is important to make sure a privacy request is created.
- Also, separately creates a snapshot of the preferences on the `ConsentRequest` object, because a `PrivacyRequest` is not necessarily created here, only if we have things to propagate.

```bash
curl -X 'PATCH' \
  'http://localhost:8080/api/v1/consent-request/con_bcec68e1-ae0f-494d-8be6-a07d296083bb/preferences' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "code": "1234",
  "consent": [
    {
      "data_use": "improve",
      "opt_in": true,
      "has_gpc_flag": false,
      "conflicts_with_gpc": false
    }
  ],
  "executable_options": [
    {
      "data_use": "improve",
      "executable": true
    }
  ]
}'
```

## Get previously-saved consent preferences
- Use your consent request id from earlier.  The code can be anything when identify verification is turned off.

```bash
curl -X 'POST' \
  'http://localhost:8080/api/v1/consent-request/con_bcec68e1-ae0f-494d-8be6-a07d296083bb/verify' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "code": "1234"
}'
```


## SQL Queries for double checking migration


### View created privacy request

```sql
fides=# select id, status, consent_preferences from privacyrequest;
                    id                    |  status  |                                                     consent_preferences                                                     
------------------------------------------+----------+-----------------------------------------------------------------------------------------------------------------------------
 pri_b926282f-9679-40d5-9d03-b239ddf6207b | complete | [{"opt_in": true, "data_use": "improve", "has_gpc_flag": false, "conflicts_with_gpc": false, "data_use_description": null}]
(1 row)
```

### View updated consent request 

```sql
fides=# select id, preferences, privacy_request_id from consentrequest;
                    id                    |                                                         preferences                                                         |            privacy_request_id            
------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------+------------------------------------------
 con_03c926d1-6840-43ba-8757-d5f06590ddf9 |                                                                                                                             | 
 con_bcec68e1-ae0f-494d-8be6-a07d296083bb | [{"opt_in": true, "data_use": "improve", "has_gpc_flag": false, "conflicts_with_gpc": false, "data_use_description": null}] | pri_b926282f-9679-40d5-9d03-b239ddf6207b

```

### View Consent resource
```sql
fides=# select id, data_use, opt_in from consent;
                    id                    | data_use | opt_in 
------------------------------------------+----------+--------
 con_2573d315-b6f8-42e2-9018-f18b4322b71c | improve  | t
```

**Comment by ThomasLaPiana** [comment]
@pattisdr amazing testing directions/context, thank you so much! I'll get these translated into the script and hopefully we'll be off to the races 🙂

**Comment by ThomasLaPiana** [comment]
using the `steps to confirm`, I got everything passing locally

File: src/fides/api/api/v1/endpoints/privacy_request_endpoints.py [hunk 1]
```diff
@@ -557,6 +557,8 @@ def get_request_status(
 ) -> Union[StreamingResponse, AbstractPage[PrivacyRequest]]:
     """Returns PrivacyRequest information. Supports a variety of optional query params.
 
+    TODO: Why is request_id a query param instead of a path param?
```
Comment (pattisdr) [inline src/fides/api/api/v1/endpoints/privacy_request_endpoints.py:560]: I believe this is the endpoint that populates the privacy requests page, so by default we return a paginated list of privacy requests.  There are several UI filters on that page, one of them allows you to filter directly by request id.

**Review by pattisdr — COMMENTED** [review]
This is looking great @ThomasLaPiana, I reviewed and have been testing this evening.  I appreciate the extra effort on the migration test script to build confidence here: I'll test a little more in the AM.  None of the comments below affect the migration itself.

> Also, what did you mean by "migrating live"? I'm not sure there's any other way? 

I just meant the application isn't up and running, no one is writing or reading from the db while this migration is performed. Just double checking, given the potential size of consent/consent request/privacy request.

File: src/fides/api/schemas/privacy_request.py [hunk 2]
```diff
@@ -62,7 +65,10 @@ class Consent(FidesSchema):
 
 
 class ConsentReport(Consent):
-    """Schema for reporting Consent requests."""
+    """
+    Schema for reporting Consent requests.
+    TODO: What is a consent report
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/schemas/privacy_request.py:70]: Sorry I should've thought of a better term here...I think `Question for Reviewers` is more what I was thinking!

This code is _dense_, and I want to add a few more breadcrumbs for the next person that comes along, so I threw some TODO hoping any reviewers could help me add definitions as part of the merge!

File: src/fides/api/schemas/privacy_request.py [hunk 3]
```diff
@@ -52,7 +52,10 @@ class Config:
 
 
 class Consent(FidesSchema):
-    """Schema for consent."""
+    """
+    Schema for consent.
+    TODO: What is consent in this situation? Are these hardcoded values that come from the frontend implementation?
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply src/fides/api/schemas/privacy_request.py:57]: Thank you for the context!

**Review by pattisdr — APPROVED** [review]

**Review by pattisdr — CHANGES_REQUESTED** [review]
Going back through the other PR, there's a comment about the RuleTarget table?

https://github.com/ethyca/fides/pull/3933/files#r1309338648

**Review by pattisdr — APPROVED** [review]
OK! take a look at my RuleTarget work please before merge!

**Comment by ThomasLaPiana** [comment]
@pattisdr verified the changes! Thank you!

**Comment by ThomasLaPiana** [comment]
failure is a flaky test, merging

**Pull request merged**